core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java

```diff
@@ -72,30 +74,35 @@ private boolean refill() throws IOException {
   }
 
   @Override
-  public int read() throws IOException {
+  public synchronized int read() throws IOException {
     if (!refill()) {
       return -1;
     }
     return byteBuffer.get() & 0xFF;
   }
 
   @Override
-  public int read(byte[] b, int offset, int len) throws IOException {
+  public synchronized int read(byte[] b, int offset, int len) throws IOException {
     if (!refill()) {
       return -1;
     }
+    if ((offset | len | (offset + len) | (b.length - (offset + len))) < 0) {
```

Comment:
@srowen I think this check is correct. It checks if any of these values is negative.

> offset and len can't be negative and their sum shouldn't exceed the array length.

It's possible. `offset` and `len` are set by the caller.

@sitalkedia Could you move the argument check before `refill`? If the argument is invalid, we should always throw an exception instead of returning `-1`.